File: web/src/utils/wind.ts
/**
 * Seeded wind animation utilities shared by layout and radar chart.
 *
 * Each branch gets a deterministic speed and phase derived from its index,
 * producing varied but repeatable organic sway. Y-axis is damped for realism.
 */

// Wind amplitude and speed range
export const WIND_BRANCH_AMP = 6
export const WIND_TWIG_AMP = 10
export const WIND_PULSE = 0.04
export const WIND_MIN = 0.35
export const WIND_MAX = 0.7
export const WIND_FOCUS_BRANCH_SCALE = 0.7
export const WIND_FOCUS_TWIG_SCALE = 0.85
export const WIND_Y_DAMPING = 0.6
export const WIND_FLUTTER_SCALE = 0.18

export function seeded(seed: number, salt: number): number {
  const v = Math.sin(seed * salt) * 43758.5453
  return v - Math.floor(v)
}

export function lerp(a: number, b: number, t: number): number {
  return a + (b - a) * t
}

export function clamp(v: number, min: number, max: number): number {
  return Math.min(Math.max(v, min), max)
}

/**
 * Compute seeded wind offset for a branch at a given time (seconds).
 * Returns {x, y} in pixel space with Y damped to 60%.
 */
export function branchWindOffset(
  index: number,
  time: number,
  amplitude = WIND_BRANCH_AMP,
): { x: number; y: number } {
  const seed = 97 + index * 41
  const speed = lerp(WIND_MIN, WIND_MAX, seeded(seed, 13.7))
  const phase = seeded(seed, 23.1) * Math.PI * 2
  return {
    x: Math.sin(time * speed + phase) * amplitude,
    y: Math.cos(time * speed * 0.8 + phase) * amplitude * WIND_Y_DAMPING,
  }
}

File: ios/Trunk/Utils/Wind.swift
//
//  Wind.swift
//  Trunk
//
//  Shared seeded wind animation utilities.
//  Used by TreeCanvasView, BranchView, and RadarChartView for consistent
//  organic sway. Each node gets a deterministic speed and phase derived
//  from its seed, producing varied but repeatable motion.
//

import CoreGraphics
import Foundation

enum Wind {
    static let branchAmplitude: CGFloat = 6.0
    static let twigAmplitude: CGFloat = 8.0
    static let yDamping: Double = 0.6
    static let minSpeed: Double = 0.35
    static let maxSpeed: Double = 0.7

    /// Seeded pseudo-random value in [0, 1) matching the web implementation.
    static func seeded(seed: Int, salt: Double) -> Double {
        let v = sin(Double(seed) * salt) * 43758.5453
        return v - floor(v)
    }

    static func lerp(_ a: Double, _ b: Double, _ t: Double) -> Double {
        a + (b - a) * t
    }

    /// Wind offset for a branch node at a given time.
    /// Seed formula: 97 + index * 41 (matches web layout.ts).
    static func branchOffset(index: Int, time: Double, amplitude: CGFloat = branchAmplitude) -> CGPoint {
        let seed = 97 + index * 41
        let speed = lerp(minSpeed, maxSpeed, seeded(seed: seed, salt: 13.7))
        let phase = seeded(seed: seed, salt: 23.1) * Double.pi * 2
        let x = sin(time * speed + phase) * Double(amplitude)
        let y = cos(time * speed * 0.8 + phase) * Double(amplitude) * yDamping
        return CGPoint(x: x, y: y)
    }

    /// Wind offset for a twig node within a branch.
    /// Seed formula: 131 + branchIndex * 71 + twigIndex * 17 (matches web layout.ts).
    static func twigOffset(branchIndex: Int, twigIndex: Int, time: Double, amplitude: CGFloat = twigAmplitude) -> CGPoint {
        let seed = 131 + branchIndex * 71 + twigIndex * 17
        let speed = lerp(minSpeed, maxSpeed, seeded(seed: seed, salt: 17.9))
        let phase = seeded(seed: seed, salt: 29.3) * Double.pi * 2
        let x = sin(time * speed + phase) * Double(amplitude)
        let y = cos(time * speed * 0.8 + phase) * Double(amplitude) * yDamping
        return CGPoint(x: x, y: y)
    }
}
